src/DurableTask.SqlServer/Scripts/logic.sql

```diff
@@ -1252,3 +1266,138 @@ BEGIN
     VALUES (@SemanticVersion)
 END
 GO
+
+
+CREATE OR ALTER PROCEDURE dt._RewindInstance
+    @InstanceID varchar(100),
+    @Reason varchar(max) = NULL
+AS
+BEGIN
+    DECLARE @TaskHub varchar(50) = dt.CurrentTaskHub()
+
+    -- *** IMPORTANT ***
+    -- To prevent deadlocks, it is important to maintain consistent table access
+    -- order across all stored procedures that execute within a transaction.
+    -- Table order for this sproc: Instances --> (History --> Payloads --> History), NewEvents)  
+
+    DECLARE @existingStatus varchar(30)
+    DECLARE @executionID varchar(50)
+    
+    SELECT TOP 1 @existingStatus = existing.[RuntimeStatus], @executionID = existing.[ExecutionID]
+    FROM Instances existing WITH (HOLDLOCK)
+    WHERE [TaskHub] = @TaskHub AND [InstanceID] = @InstanceID
+
+    -- Instance IDs can be overwritten only if the orchestration is in a terminal state
+    IF @existingStatus NOT IN ('Failed')
+    BEGIN
+        DECLARE @msg nvarchar(4000) = FORMATMESSAGE('Cannot rewing instance with ID ''%s'' because it is not in a ''Failed'' state, but in ''%s'' state.', @InstanceID, @existingStatus);
+        THROW 50001, @msg, 1;
+    END
+    
+    DECLARE @eventsInFailure TABLE (
+        [SequenceNumber] bigint NULL,
+        [EventType] varchar(40) NULL,
+        [TaskID] int NULL,
+        [DataPayloadID] uniqueidentifier NULL)
+    
+    -- List all events related to failures (ie TaskScheduled/TaskFailed and SubOrchestrationInstanceStarted/SubOrchestrationInstanceFailed couples)
+    INSERT INTO @eventsInFailure
+    SELECT h.[SequenceNumber], h.[EventType], h.[TaskID], h.[DataPayloadID]
+    FROM History h
+    WHERE h.[TaskHub] = @TaskHub AND h.[InstanceID] = @InstanceID
+      AND (h.[EventType] IN ('TaskFailed', 'SubOrchestrationInstanceFailed')
+           OR (h.[EventType] IN ('TaskScheduled', 'SubOrchestrationInstanceStarted') AND EXISTS (SELECT 1
+                                                           FROM History f
+                                                           WHERE f.[TaskHub] = @TaskHub AND f.[InstanceID] = @InstanceID  AND f.[TaskID] = h.[TaskID] AND f.[EventType] = CASE WHEN h.[EventType] = 'TaskScheduled' THEN 'TaskFailed' ELSE 'SubOrchestrationInstanceFailed' END)))
```

Comment:
Making a note to myself to spend more time reviewing what this query is doing.

Reply:
Just as an initial help, it retrieves all `TaskFailed` and `SubOrchestrationInstanceFailed`, and their corresponding `TaskScheduled` and `SubOrchestrationInstanceStarted`.

Reply:
Thanks. I also took the liberty of reformatting this a bit to reduce horizontal scrolling, making it easier for me to read. :)